Refactor capabilities

xsebek opened the pull request:
* add `CExecute Const`
  * remove all capabilities for constants
  * update YAML to mention each capability for constant explicitly
* uses [`generic-data`](https://hackage.haskell.org/package/generic-data-1.1.0.0/docs/Generic-Data.html#t:FiniteEnumeration) package to get `Enum` and `Bounded`
  ```Haskell
  deriving (Enum, Bounded) via (FiniteEnumeration Capability)
  ```
* closes #1548
* closes #2018

xsebek commented on src/Swarm/Language/Capability.hs:
```diff
@@ -136,27 +59,8 @@ data Capability
-  | -- | Capability to obtain absolute time, namely via the `time` command.
-    CTimeabs
-  | -- | Capability to utilize relative passage of time, namely via the `wait` command.
-    --   This is strictly weaker than "CTimeAbs".
```
I would like to get rid of these confusing capabilities that say they do more, when it is the entities that do the work.

The `CTimerel` capability _is not strictly weaker itself_, this logic is in entities providing it along with `CTimeabs`.

The idea would be to specify all the commands that entities want to enable in the capability list.

@byorgey @kostmo what do you think? 🤔

byorgey replied:
We should just change `CTimeabs` into `CExecute Time` and `CTimerel` into `CExecute Wait`, I think.  See my other comment below.

xsebek commented on src/Swarm/Language/Capability.hs:
```diff
@@ -253,39 +169,17 @@ constCaps = \case
   Self -> Just CWhoami
-  Swap -> Just CSwap
-  Atomic -> Just CAtomic
-  Instant -> Just CGod
```
I think this change will significantly simplify keeping `execCaps` up to date. 🙂 

As it is now clear to see we had `Instant` and `Use` in the wrong sections.

byorgey commented on src/Swarm/Language/Capability.hs:
```diff
@@ -136,27 +59,8 @@ data Capability
-  | -- | Execute the `meet` and `meetAll` commands.
-    CMeet
   | -- | Capability to introspect and see its own name
     CWhoami
```
Oh, never mind, I see this capability is for both `whoami` and `self` commands.

xsebek replied:
Yes, but perhaps we can have two self explanatory capabilities or have self implied by whoami.

byorgey replied:
Sure, it should always be OK to (1) split out capabilities like this into individual capabilities for each command, then (2) edit any devices which currently provide the old combined capability so that they now provide both.  The only reason not to do that before was to prevent a proliferation of lots of capabilities, but now that we have `CExecute` that's not really an issue.

In fact, it's probably a good idea to just split all such capabilities since it gives more fine-grained control for scenario authors.  For example, the `time` capability used to allow running both `time` and `wait` commands, and we split it into `timeabs` and `timerel` so that we could have scenarios that allow you to run `wait` but not `time`.

xsebek replied:
Done! That took a lot of effort, but our test coverage helped a lot. 😄

byorgey commented on src/Swarm/TUI/Controller.hs:
```diff
@@ -79,7 +79,9 @@ import Swarm.Game.Robot
 import Swarm.Game.Step (finishGameTick, gameTick)
-import Swarm.Language.Capability (Capability (CDebug, CMake))
+
+-- See Note [liftA2 re-export from Prelude]
```
I think this comment got moved inadvertently by `fourmolu`, perhaps?

xsebek replied:
Yeah it was some fail, I got two restyled PRs in succession. Will take a look, maybe it’s something we could report to fourmolu.

xsebek commented on src/Swarm/Language/Capability.hs:
```diff
@@ -166,17 +70,53 @@ data Capability
+    , ("debug", CDebug)
+    , ("heavy robot move", CMoveHeavy)
+    , ("handle input", CHandleInput)
+    , ("god", CGod)
```
I think we can - pattern match in case, construct a map, then invert the map. 🤔

byorgey commented:
@xsebek what's the status of this PR?  Can I help?

xsebek commented:
@byorgey I seem to have forgotten about this; thanks for bringing it up.

I think this is a worthwhile cleanup but can not say when I will be able to get around to rebasing this on the current main.

xsebek commented:
@kostmo @byorgey, I rebased this old PR and split capabilities to explicitly mention the `Const`, as @byorgey suggested.

I tried to use `universe`, but as far as I can tell, it does not provide generic deriving. So I went with `generic-data`, which is very lightweight:
```
$ cabal-plan diff --plan-json ../plan_orig.json --relative .

Package versions
~~~~~~~~~~~~~~~~

+ap-normalize-0.1.0.1 lib
+generic-data-1.1.0.0 lib
+show-combinators-0.2.0.0 lib
```

xsebek commented on src/swarm-scenario/Swarm/Game/Device.hs:
```diff
@@ -71,7 +71,7 @@ data CapabilityCost e = CapabilityCost
 -- Otherwise, parse as a Map from capabilities to ingredients.
 instance (FromJSON e) => FromJSON (SingleEntityCapabilities e) where
   parseJSON x =
-    simpleList <|> (Capabilities <$> costMap)
```
@kostmo I swapped these to get better errors, as they only told me that the wrong capability is not an object.

I don't know if this can be improved much. Maybe the costly capabilities could be under a separate key.

xsebek commented:
Here is the updated cheatsheet:
```sh
cabal run swarm-docs -O0 -- cheatsheet --capabilities
```

<details>
  <summary>Expand table</summary>


| Name              | Commands                                                            | Entities                              |
|-------------------|---------------------------------------------------------------------|---------------------------------------|
| wait              | `wait`                                                              | clock, hourglass, rolex               |
| selfdestruct      | `selfdestruct`                                                      | detonator                             |
| move              | `move`                                                              | treads, tank treads                   |
| backup            | `backup`                                                            | tape drive                            |
| volume            | `volume`                                                            |                                       |
| path              | `path`                                                              |                                       |
| push              | `push`                                                              | dozer blade                           |
| stride            | `stride`                                                            |                                       |
| turn              | `turn`                                                              | treads, tank treads                   |
| grab              | `grab`                                                              | grabber, fast grabber, harvester      |
| harvest           | `harvest`                                                           | harvester                             |
| sow               | `sow`                                                               |                                       |
| ignite            | `ignite`                                                            |                                       |
| place             | `place`                                                             | grabber, fast grabber, harvester      |
| ping              | `ping`                                                              |                                       |
| give              | `give`                                                              | grabber, fast grabber                 |
| equip             | `equip`                                                             | welder                                |
| unequip           | `unequip`                                                           | welder                                |
| make              | `make`                                                              | workbench                             |
| count             | `count`                                                             | counter                               |
| drill             | `drill`                                                             | drill, metal drill                    |
| build             | `build`                                                             | 3D printer                            |
| salvage           | `salvage`                                                           | toolkit                               |
| reprogram         | `reprogram`                                                         | flash memory                          |
| listen            | `listen`                                                            | hearing aid                           |
| log               | `log`                                                               | logger                                |
| view              | `view`                                                              |                                       |
| appear            | `appear`                                                            | beaglepuss                            |
| halt              | `halt`                                                              | halting oracle                        |
| time              | `time`                                                              | clock, rolex                          |
| scout             | `scout`                                                             | binoculars                            |
| whereami          | `whereami`                                                          | GPS receiver                          |
| waypoint          | `waypoint`                                                          |                                       |
| structure         | `structure`                                                         | blueprint                             |
| floorplan         | `floorplan`                                                         | blueprint                             |
| hastag            | `hastag`                                                            |                                       |
| tagmembers        | `tagmembers`                                                        |                                       |
| detect            | `detect`                                                            |                                       |
| resonate          | `resonate`                                                          |                                       |
| density           | `density`                                                           |                                       |
| sniff             | `sniff`                                                             | olfactometer                          |
| chirp             | `chirp`                                                             |                                       |
| watch             | `watch`                                                             | rolex                                 |
| heading           | `heading`                                                           | compass                               |
| blocked           | `blocked`                                                           | scanner                               |
| scan              | `scan`                                                              | scanner                               |
| upload            | `upload`                                                            | scanner                               |
| ishere            | `ishere`                                                            | scanner                               |
| isempty           | `isempty`                                                           | scanner                               |
| self              | `self`                                                              | mirror                                |
| meet              | `meet`                                                              | antenna                               |
| meetall           | `meetall`                                                           | antenna                               |
| whoami            | `whoami`                                                            | mirror                                |
| random            | `random`                                                            | bitcoin                               |
| if                | `if`                                                                | branch predictor                      |
| try               | `try`                                                               | net                                   |
| not               | `not`                                                               | lodestone                             |
| neg               | `-`                                                                 | calculator, ADT calculator, hyperloop |
| ==                | `==`                                                                | comparator                            |
| !=                | `!=`                                                                | comparator                            |
| <                 | `<`                                                                 | comparator                            |
| >                 | `>`                                                                 | comparator                            |
| <=                | `<=`                                                                | comparator                            |
| >=                | `>=`                                                                | comparator                            |
| \|\|              | `\|\|`                                                              | branch predictor                      |
| &&                | `&&`                                                                | branch predictor                      |
| +                 | `+`                                                                 | calculator, ADT calculator, hyperloop |
| -                 | `-`                                                                 | calculator, ADT calculator, hyperloop |
| *                 | `*`                                                                 | calculator, ADT calculator, hyperloop |
| /                 | `/`                                                                 | calculator, ADT calculator, hyperloop |
| ^                 | `^`                                                                 | calculator, ADT calculator, hyperloop |
| format            | `format`                                                            | linotype, string                      |
| ++                | `++`                                                                | Elmer's glue, string                  |
| chars             | `chars`                                                             | caliper, string                       |
| split             | `split`                                                             | wedge, string                         |
| charat            | `charat`                                                            | decoder ring                          |
| tochar            | `tochar`                                                            | decoder ring                          |
| swap              | `swap`                                                              | fast grabber                          |
| atomic            | `atomic`                                                            | fast grabber, rubber band             |
| key               | `key`                                                               | keyboard                              |
| installkeyhandler | `installkeyhandler`                                                 | keyboard                              |
| teleport          | `teleport`                                                          | infinite improbability drive          |
| knows             | `knows`                                                             |                                       |
| power             |                                                                     | solar panel                           |
| move heavy robot  |                                                                     | tank treads                           |
| float             |                                                                     | boat                                  |
| orient            |                                                                     | compass                               |
| env               |                                                                     | dictionary                            |
| lambda            |                                                                     | lambda                                |
| recursion         |                                                                     | strange loop                          |
| sum               | `inl`, `inr`, `case`                                                | ADT calculator, hyperloop             |
| prod              | `fst`, `snd`                                                        | ADT calculator, hyperloop             |
| record            |                                                                     | victrola                              |
| debug             |                                                                     | tweezers                              |
| rectype           |                                                                     | hyperloop                             |
| god               | `create`, `surveil`, `instant`, `as`, `robotnamed`, `robotnumbered` |                                       |

</details>

byorgey commented on src/swarm-lang/Swarm/Language/Capability.hs:
```diff
@@ -229,124 +124,23 @@ constCaps = \case
-  -- Some more constants which *ought* to have their own capability but
-  -- currently don't.
-  View -> Nothing -- TODO: #17 should require equipping an antenna
-  Knows -> Nothing
```
`view` and `knows` used to require no capabilities, but you (probably inadvertently) removed them, so now it is impossible to execute either of them.  For now we should put them back along with `run` and `use`.

xsebek replied:
Ups, I should learn to move code up instead of Ctrl+X. 😅

byorgey commented on src/swarm-doc/Swarm/Doc/Wiki/Cheatsheet.hs:
```diff
@@ -188,7 +188,11 @@ capabilityTable a em cs = T.unlines $ header <> map (listToRow mw) capabilityRow
-capabilityPage a em = capabilityTable a em enumerate
+capabilityPage a em = capabilityTable a em $ filter usedCapabily enumerate
+ where
+  usedCapabily c = case c of
```
`usedCapability`?

byorgey commented on src/swarm-lang/Swarm/Language/Capability.hs:
```diff
@@ -181,20 +64,32 @@ data Capability
+capabilityName = \case
+  CExecute con -> case con of
+    Neg -> "neg"
+    Not -> "not"
```
Why do we need a special case for `Not`?  (I think I understand why we need a special case for `Neg`: because its syntax is `-` which is identical to the syntax for `Sub` --- right?)

xsebek replied:
We do `not`. It was a leftover from a intermediate state, where I changed it because it originally require `negation` capability which was confusing.

byorgey commented on src/swarm-lang/Swarm/Language/Capability.hs:
```diff
@@ -181,20 +64,32 @@ data Capability
+  CExecute con -> case con of
+    Neg -> "neg"
+    Not -> "not"
+    _ -> syntax $ constInfo con
```
Are we sure `syntax . constInfo` always returns only lowercase?

xsebek replied:
Is that not the same code used to normally parse `Const`? The implementation does use `tShow` so we can be about as sure as possible without writing a newtype.

byorgey reviewed (approved):
You're right, that did end up being a lot of work.  One of those things that seems like it will be simple/straightforward and ends up being more work than you thought.  Thanks for doing it though, I think this is a big improvement!
